lib/Backend/ServerThreadContext.cpp

```diff
@@ -4,19 +4,144 @@
 
 #include "Backend.h"
+
 #if ENABLE_OOP_NATIVE_CODEGEN
 #include "JITServer/JITServer.h"
-#endif
+
+CriticalSection PageAllocatorPool::cs;
+PageAllocatorPool* PageAllocatorPool::Instance = nullptr;
+unsigned long long PageAllocatorPool::timerWating;
+
+PageAllocatorPool::PageAllocatorPool()
+    :pageAllocators(&NoThrowHeapAllocator::Instance),
+    activePageAllocatorCount(0)
+{
+    idleCleanupTimer = CreateWaitableTimerEx(NULL, L"JITServerIdle", 0/*auto reset*/, TIMER_ALL_ACCESS);
+}
+
+PageAllocatorPool::~PageAllocatorPool()
+{
+    RemoteAll();
+}
+
+void PageAllocatorPool::Initialize()
+{
+    Instance = HeapNewNoThrow(PageAllocatorPool);
+    if (Instance == nullptr)
+    {
+        Js::Throw::FatalInternalError();
+    }
+}
+void PageAllocatorPool::Shutdown()
+{
+    AutoCriticalSection autoCS(&cs);
+    HeapDelete(Instance);
+    Instance = nullptr;
+}
+void PageAllocatorPool::RemoteAll()
+{
+    AutoCriticalSection autoCS(&cs);
+    while (!pageAllocators.Empty())
+    {
+        HeapDelete(pageAllocators.Pop());
+    }
+}
+
+void PageAllocatorPool::IncreaseActiveCount()
+{
+    InterlockedIncrement(&activePageAllocatorCount);
+}
+void PageAllocatorPool::DecreaseActiveCount()
+{
+    unsigned long long currentActiveCount = InterlockedDecrement(&PageAllocatorPool::Instance->activePageAllocatorCount);
+    if (currentActiveCount == 0)
+    {
+        PageAllocatorPool::IdleCleanup();
+    }
+}
+
+PageAllocator* PageAllocatorPool::GetPageAllocator()
+{
+    AutoCriticalSection autoCS(&cs);
+    PageAllocator* pageAllocator = nullptr;
+    if (pageAllocators.Count() > 0)
+    {
+        // TODO: OOP JIT, select the page allocator with right count of free pages
+        // base on some heuristic
+        pageAllocator = this->pageAllocators.Pop();
+    }
+    else
+    {
+        pageAllocator = HeapNew(PageAllocator, nullptr, Js::Configuration::Global.flags, PageAllocatorType_BGJIT,
+            AutoSystemInfo::Data.IsLowMemoryProcess() ? PageAllocator::DefaultLowMaxFreePageCount : PageAllocator::DefaultMaxFreePageCount);
+    }
+
+    return pageAllocator;
+
+}
+void PageAllocatorPool::ReturnPageAllocator(PageAllocator* pageAllocator)
+{
+    AutoCriticalSection autoCS(&cs);
+
+    if (!this->pageAllocators.PrependNoThrow(&HeapAllocator::Instance, pageAllocator))
+    {
+        HeapDelete(pageAllocator);
+    }
+}
+
+void PageAllocatorPool::IdleCleanup()
+{
+    AutoCriticalSection autoCS(&cs);
+    __analysis_assume(Instance);
+    if (timerWating == 0)
+    {
+        LARGE_INTEGER liDueTime;
+        liDueTime.QuadPart = -10 * 10000000LL; // wait for 10 seconds to do the cleanup
+        if (!SetWaitableTimer(Instance->idleCleanupTimer, &liDueTime, 0, IdleCleanupRoutine, NULL, 0))
+        {
+            if (Instance)
+            {
+                Instance->RemoteAll();
+            }
+        }
+        InterlockedExchange(&timerWating, 1);
```

Comment:
removed

---
In reply to: [83754825](https://github.com/Microsoft/ChakraCore/pull/1742#discussion_r83754825) [](ancestors = 83754825)